src/instructions/segmentation.rs

```diff
@@ -1,210 +1,278 @@
-/// Reload data segment register.
-///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid data segment descriptor.
-#[inline]
-pub unsafe fn load_ds(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov ds, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+macro_rules! segment_impl {
+    ($type:ty, $name:literal, $asm_get:ident, $asm_load:ident) => {
+        impl Segment for $type {
+            get_reg_impl!($name, $asm_get);
+
+            unsafe fn set_reg(sel: SegmentSelector) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("mov ", $name, ", {0:x}"), in(reg) sel.0, options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_ds(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_load(sel.0);
+            }
+        }
+    };
 }
 
-/// Reload es segment register.
-///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid extra segment descriptor.
-#[inline]
-pub unsafe fn load_es(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov es, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+macro_rules! segment64_impl {
+    ($type:ty, $name:literal, $base:ty, $asm_rd:ident, $asm_wr:ident) => {
+        impl Segment64 for $type {
+            const BASE: Msr = <$base>::MSR;
+            unsafe fn read_base() -> VirtAddr {
+                #[cfg(feature = "inline_asm")]
+                {
+                    let val: u64;
+                    asm!(concat!("rd", $name, "base {}"), out(reg) val, options(nomem, nostack, preserves_flags));
+                    VirtAddr::new_unsafe(val)
+                }
+                #[cfg(not(feature = "inline_asm"))]
+                VirtAddr::new_unsafe(crate::asm::$asm_rd())
+            }
+
+            unsafe fn write_base(base: VirtAddr) {
+                #[cfg(feature = "inline_asm")]
+                asm!(concat!("wr", $name, "base {}"), in(reg) base.as_u64(), options(nostack, preserves_flags));
 
-    #[cfg(not(feature = "inline_asm"))]
-    crate::asm::x86_64_asm_load_es(sel.0);
+                #[cfg(not(feature = "inline_asm"))]
+                crate::asm::$asm_wr(base.as_u64());
+            }
+        }
+    };
 }
 
-/// Reload fs segment register.
+/// Code Segment
 ///
-/// ## Safety
-///
-/// This function is unsafe because the caller must ensure that `sel`
-/// is a valid fs segment descriptor.
-#[inline]
-pub unsafe fn load_fs(sel: SegmentSelector) {
-    #[cfg(feature = "inline_asm")]
-    asm!("mov fs, {0:x}", in(reg) sel.0, options(nostack, preserves_flags));
+/// The segment base and limit are unused in 64-bit mode. Only the L (long), D
+/// (default operation size), and DPL (descriptor privilege-level) fields of the
+/// descriptor are recognized. So chaning the segment register can be used to
```

Comment:
```suggestion
/// descriptor are recognized. So changing the segment register can be used to
```
or

```suggestion
/// descriptor are recognized. So the segment register can be used to
```

?

Reply:
Fixed in https://github.com/rust-osdev/x86_64/pull/258/commits/a81714bc90184a5d60d5ec56aefe0c08ef57b83c with the first suggestion.